### PR TITLE
Add RandomGaussianBlur with instance-level gaussian kernel generation

### PR DESCRIPTION
**Pull request by juliendenize**
#### Changes

Fixes a comment I put in #691  

The current implementation of `RandomGaussianBlur` is random in the sense of only some instances are augmented given the `p` probability, with a shared fixed gaussian kernel. This differs from other existing approaches such as torchvision [`GaussianBlur`](http://pytorch.org/vision/main/generated/torchvision.transforms.GaussianBlur.html).

In Torchvision philosophy, the randomness of gaussian blur is on the sampling of the `sigma` std factor to generate the gaussian kernels. To allow reproducibility of various works that are based on torchvision such as papers on Self-Supervised Learning for example (SimCLR, MoCo, Byol, ...), just to cite this area of work, it would be a great feature to have in Kornia.

I propose a new version of `RandomGaussianBlur` that takes as an input a sigma tuple to provide bounds for uniform sampling. A filter is then generated independently for each instance following a vectorized process to not provoke a bottleneck.

I renamed the previous `RandomGaussianBlur` back to `GaussianBlur` but I can change the names to whatever you feel is best.

I implemented tests to validate the process and everything works except the gradcheck through the module but I don't know why, so I'm more than happy to receive help to debug this aspect.

#### Type of change

- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/filters/gaussian.py [hunk 1]
```diff
@@ -3,8 +3,59 @@
...
 import torch.nn as nn
 
+from kornia.testing import KORNIA_CHECK_IS_TENSOR
+
 from .filter import filter2d, filter2d_separable
-from .kernels import get_gaussian_kernel1d, get_gaussian_kernel2d
+from .kernels import get_gaussian_kernel1d, get_gaussian_kernel2d, get_n_gaussian_kernel1d, get_n_gaussian_kernel2d
+
+
+def random_gaussian_blur2d(
```
Comment (edgarriba) [inline kornia/filters/gaussian.py:12]: I don’t think we want to mix operations that involves random states in the Kornia core functionality. The main application for random states for us data augmentation, and for this it should go to the dedicated module. However, another discussion is whether to provide free functions — in the past we had, but we end up deprecating because of the overhead to maintain and not fully fitting in the full provided features from the modules api e.g. return transforms. So, I’d everything that involves random goes to Kornia.augmentation and in the form of a module. /cc @shijianjian

File: kornia/filters/kernels.py [hunk 2]
```diff
@@ -24,6 +24,21 @@ def gaussian(window_size: int, sigma: float) -> torch.Tensor:
     return gauss / gauss.sum()
 
 
+def n_gaussian(window_size: int, sigma: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/filters/kernels.py:27]: Can we just update the previous function accepting both float and tensor ? If float then convert to tensor. No need code duplication

File: kornia/augmentation/__init__.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
     ColorJiggle,
     ColorJitter,
     Denormalize,
+    GaussianBlur,
```
Comment (edgarriba) [inline kornia/augmentation/__init__.py:6]: I’d say that we provide only RandomGaussianBlur — if the needed feature is sample batches of sigma then let’s upgrade the current functionality as you did. In Kornia world, to me plane GaussianBlur should be the Module version of Kornia.filters.gaussian_blur2d /cc @shijianjian @ducha-aiki

File: kornia/filters/gaussian.py (hunk 1 above)
Comment (juliendenize) [reply kornia/filters/gaussian.py:12]: It's really bad named function, sorry about that I refractored the code several times. There is no random logic but it is the same function as `gaussian_blur2d`, however it takes a sigma tensor that could be of shape `Bx2` instead of float. Based on one other comment u made it could probably be only one function instead of one taking in float and the other one taking in a tensor.

File: kornia/filters/kernels.py (hunk 2 above)
Comment (juliendenize) [reply kornia/filters/kernels.py]: I updated the code to do that

File: kornia/augmentation/_2d/intensity/gaussian_blur.py [hunk 4]
```diff
@@ -7,7 +7,7 @@
 from kornia.filters import gaussian_blur2d
 
 
-class RandomGaussianBlur(IntensityAugmentationBase2D):
+class GaussianBlur(IntensityAugmentationBase2D):
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: what I meant in the previous comments is that this entire file/function might disappear and we keep only `RandomGaussianBlur` /cc @shijianjian

Comment (juliendenize) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: I better understand what you mean, I'll wait for you and other members decision to remove one of the two. However, if I may intervene, I still think it might be interesting to keep both as they have different purposes. Maybe clearer names for their goals should be good enough to properly manage the two in the library.

Comment (edgarriba) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: `kornia.augmentation` operators that involves randomness -- the rest in the corresponding `kornia.*` module. I'd like to keep simple with only `RandomGaussianBlur` to reduce the number of API  to also easier maintenance. Maybe we can just describe here what what you have in mind that differ the both implementation so that we can find a good single pattern.

Comment (juliendenize) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: Ok i'll resume below the two implementations 

- Current implementation takes a tuple as sigma to specify a fix std to construct a gaussian kernel with sigma[0] for the first axis and sigma[1] for the second axis. The randomness comes with the probability of each instance to be augmented or not.
- My proposed implementation takes a tuple as sigma to specify the range of random drawn sigma std. The drawn sigma is the same for both axis. The randomness comes with the probability of each instance to be augmented or not and that each instance is associated with a different kernel built upon the drawn sigma.

To have a single implementation for both, we could add a `sigma_mode` argument for the init module that is one of these two values respectively for both implementations:
- fix
- random

Comment (shijianjian) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: Can't we follow the same conventions to add the `separable`? https://github.com/kornia/kornia/blob/c3fe305c6003edecc48f0d3c92f31977fcad7935/kornia/filters/gaussian.py#L10-L14

Also, I think in general:
1. Please try not change class or function names, since it might break the backward compatibility. If a name changing is required, please add deprecation warnings?
2. If changes the default class behaviour, a user warning is required to notify the users.

Comment (juliendenize) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: I don't think separable is the right name for it as it doesn't quite describe what this PR brings. Separable allows for computing separately the axes which nothing to do with random sigma. However, I will push a version of RandomGaussianBlur with random sigma that allows for specifying separable as I empirically observed a faster computation with separate kernels.

Based on your other remarks, when we will settle on how we want the classes to be merged or if we only keep one of them, I will follow your guidance :)

Comment (shijianjian) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: I meant to have `RandomGaussianBlur` that takes `separable` as a argument. Several scenarios here:
- If `seperable=True`, it can randomly generate x sigma and y sigma, by `sigma=(a,b)`. 
- If `seperable=True`, it can randomly generate x sigma from `(a,b)` and y sigma from `(c,d)`, by `sigma=((a,b), (c, d))`. 
- If `seperable=False`, it can randomly generate sigma, by `sigma=(a,b)`. 
- If `seperable=False`, with `sigma=((a,b), (c, d))`, an error will be thrown out.

Also, if one want to fix the sigma. He may set `sigma=(a, a)` or `sigma=((a,a),(b,b))`.

Comment (edgarriba) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: @juliendenize i like the `sigma_mode` thing -- from my perspective I care about maintenance and health of the repo and would love to avoid ending to the follwing scenerio:
```python
from K.augmentation import GaussianBlur
from K.filters import GaussianBlur
```
that's why I vote for having two clear different api:
- `K.augmentation.RandomGaussianBlur` that will handle randomness over occurrency (`p` param) and `sigma` generation
- `K.filters.GaussianBlur` that will just implement the functional of `gaussian_blur` and no random implied.

As for the `separable` args -- to give a bit of context, we included the separable kernels basically to improve performance.

Comment (juliendenize) [reply kornia/augmentation/_2d/intensity/gaussian_blur.py:10]: @edgarriba @shijianjian if I understood correctly we have 3 different propositions:
- Mine: one class under `K.augmentation.RandomGaussianBlur` with `sigma_mode in ['fix', 'random']`
- @shijianjian 's: one class under `K.augmentation.RandomGaussianBlur` with the sigma mode defined by both `separable` and the input type of sigma `Union[Tuple[float,float],Tuple[Tuple[float,float],Tuple[float,float]]]`
- @edgarriba 's: one class under `K.augmentation.RandomGaussianBlur` that does what I pushed in this commit and `K.filters.GaussianBlur` that does what `K.filters.GaussianBlur2d` already does.
We can settle on one and I will do the changes accordingly, do you both have the same preference over the three choices?

File: test/augmentation/test_augmentation.py [hunk 5]
```diff
@@ -2579,14 +2580,89 @@ def build_input(channels, height, width, bs=1, row=None, device='cpu', dtype=tor
...
+
+        # test parameters
+        batch_shape = (3, 3, 5, 5)
+        kernel_size = (3, 3)
+        sigma = (1.5, 2.1)
+
+        # evaluate function gradient
+        input = torch.rand(batch_shape, device=device, dtype=dtype)
+        input = utils.tensor_to_gradcheck_var(input)  # to var
+        assert gradcheck(RandomGaussianBlur(kernel_size, sigma, "replicate", p=1.), (input,), raise_exception=True)
```
[the remark this answers is not in the document]
Comment (juliendenize) [reply test/augmentation/test_augmentation.py:2634]: @edgarriba The tests still fail at this line, however, I don't understand why, do you have any suggestions? the weird part is that the gradcheck on the function `gaussian_blur2d` works and the only thing I add is the class wrapper and the sampling of sigma that shouldn't be in the graph.

File: kornia/filters/filter.py [hunk 6]
```diff
@@ -175,7 +175,7 @@ def filter2d_separable(input: torch.Tensor,
                   [0., 5., 5., 5., 0.],
                   [0., 0., 0., 0., 0.]]]])
     """
-    out_x = filter2d(input, kernel_x.unsqueeze(0), border_type, normalized, padding)
+    out_x = filter2d(input, kernel_x.unsqueeze(1), border_type, normalized, padding)
```
Comment (edgarriba) [inline kornia/filters/filter.py:178]: why is this needed to be changed ? can we provide a test ?

File: test/augmentation/test_augmentation.py [hunk 7]
```diff
@@ -2577,16 +2577,80 @@ def build_input(channels, height, width, bs=1, row=None, device='cpu', dtype=tor
...
+
+        # test parameters
+        batch_shape = (3, 3, 5, 5)
+        kernel_size = (3, 3)
+        sigma = torch.tensor([[1.5, 2.1], [1.5, 2.1], [1.5, 2.1]])
+
+        # evaluate function gradient
+        input = torch.rand(batch_shape, device=device, dtype=dtype)
+        input = utils.tensor_to_gradcheck_var(input)  # to var
+        assert gradcheck(kornia.filters.gaussian_blur2d, (input, kernel_size, sigma, "replicate"),
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:2620]: move this test in filters as `test_gradcheck_tensor` (or similar_

File: kornia/filters/kernels.py [hunk 8]
```diff
@@ -334,27 +339,36 @@ def get_spatial_gradient_kernel3d(mode: str, order: int, device=torch.device('cp
...
     Args:
         kernel_size: filter size. It should be odd and positive.
-        sigma: gaussian standard deviation.
+        sigma: gaussian standard deviation with shape of
+          :math:`(B, 2)`.
         force_even: overrides requirement for odd kernel size.
 
     Returns:
-        1D tensor with gaussian filter coefficients.
+        torch.Tensor: gaussian filter coefficients
```
Comment (edgarriba) [inline kornia/filters/kernels.py:356]:
```suggestion
       gaussian filter coefficients
```
type is automatically added by sphinx

File: kornia/filters/kernels.py [hunk 9]
```diff
@@ -423,42 +437,57 @@ def get_gaussian_erf_kernel1d(kernel_size: int, sigma: float, force_even: bool =
...
+                 [6.9625e-08, 2.8089e-05, 2.0755e-04, 2.8089e-05, 6.9625e-08]]])
     """
     if not isinstance(kernel_size, tuple) or len(kernel_size) != 2:
         raise TypeError(f"kernel_size must be a tuple of length two. Got {kernel_size}")
-    if not isinstance(sigma, tuple) or len(sigma) != 2:
-        raise TypeError(f"sigma must be a tuple of length two. Got {sigma}")
+    if isinstance(sigma, tuple):
+        sigma = torch.tensor(sigma).unsqueeze(0)
+    if isinstance(sigma, torch.Tensor) and sigma.shape[-1] != 2:
+        raise TypeError(f"sigma must be a tensor of shape (1, 2) or (B,2) two. Got {sigma}")
```
Comment (edgarriba) [inline kornia/filters/kernels.py:483]: use new test API: `KORNIA_CHECK_SHAPE(sigma, ["B", "2"])`

File: test/augmentation/test_augmentation.py [hunk 10]
```diff
@@ -2577,16 +2577,80 @@ def build_input(channels, height, width, bs=1, row=None, device='cpu', dtype=tor
...
+        actual = aug(input)
+        assert_close(actual, actual)
+
+    def test_gradcheck(self, device, dtype):
+        torch.manual_seed(0)
+
+        # test parameters
+        batch_shape = (3, 3, 5, 5)
+        kernel_size = (3, 3)
+        sigma = torch.tensor([[1.5, 2.1], [1.5, 2.1], [1.5, 2.1]])
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:2615]:
```suggestion
        sigma = torch.tensor([[1.5, 2.1], [1.5, 2.1], [1.5, 2.1]], device=device, dtype=dtype)
```

File: test/augmentation/test_augmentation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply test/augmentation/test_augmentation.py:2634]: I'm trying in local and cannot see what's going on. Maybe something related to the samplers ? @shijianjian

**Comment by edgarriba** [comment]
I would like to merge and include the release next week

File: kornia/filters/filter.py (hunk 6 above)
Comment (juliendenize) [reply kornia/filters/filter.py:178]: It is needed because batch size should be the first dimension but before the function, despite what the documentation specified, expected the batch size to be 1

Comment (juliendenize) [reply kornia/filters/filter.py:178]: If you look at the previous code, the function was called with `kernel_size[None]` meaning that the batch size was not considered which is no longer the case with this update.

File: test/augmentation/test_augmentation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (juliendenize) [reply test/augmentation/test_augmentation.py:2612]: @edgarriba @shijianjian It's been a while since I've worked on this PR but today I noticed something strange if I put a deterministic sigma (sigma_min = sigma_max) the test works.

I really don't get what's going on, do you have some insights ?

File: kornia/filters/kernels.py [hunk 11]
```diff
@@ -13,15 +15,20 @@ def normalize_kernel2d(input: torch.Tensor) -> torch.Tensor:
     return input / (norm.unsqueeze(-1).unsqueeze(-1))
 
 
-def gaussian(window_size: int, sigma: float) -> torch.Tensor:
-    device, dtype = None, None
-    if isinstance(sigma, torch.Tensor):
+def gaussian(window_size: int, sigma: Union[float, torch.Tensor]) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (juliendenize) [reply kornia/filters/kernels.py:18]: @edgarriba the problem is located where `sigma` is generated, if I put the line of code in the function `apply_transform`
```
 sigma = torch.tensor([[2.0085, 2.0085], [2.0794, 2.0794], [1.6040, 1.6040]], device=input.device, dtype=input.dtype)
```

instead of 
```
 sigma = params["sigma"].to(device=input.device, dtype=input.dtype).unsqueeze(-1).expand(-1, 2)
 ```

it works.

[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/filters/kernels.py:18]: i see -- I'd suggest to isolate in a separated PR to upgrade that function to use Tensor

**Comment by juliendenize** [comment]
Hi @edgarriba , 

I've rebased on master the code, from what I found the gradcheck error might be related to internal PyTorch functions. https://github.com/pytorch/pytorch/issues/78346

In the testing file of the class, I test grad check with `sigma_min = sigma_max` which passes, and I marked the testing of non-deterministic sigma as potential failure.

The function gradcheck (that does not sample sigma) properly works so, it's one more hint that it's related to PyTorch and not a failure on my side.

I hope that you'll find this PR interesting and almost ready to merge on master (minor things you might want to change) and that the tests are enough to show that the new GaussianBlur works appropriately.

I'm waiting for your feedback :)

**Comment by juliendenize** [comment]
Hi @edgarriba   ,

I've rebased on master the code, from what I found the gradcheck error might be related to how gradcheck basically works. The random aspect of sampling sigma makes it fail. If you look at other gradchecks in the testing file, the randomness is basically deactivated (Like ColorJitter or ColorJiggle put saturation, hue, contrast, etc at 0)

In the testing file of the class, I test gradcheck with `sigma_min = sigma_max` which passes, and I marked the testing of non-deterministic sigma as a potential failure but it can just be removed from the testing file.

In the testing file of the functional gaussian_blur, it passes.

So if you agree, I would basically say that there was no bug at all but just an expected behavior that I didn't know about.

I hope that you'll find this PR interesting and almost ready to merge on master (minor things you might want to change) and that the tests are enough to show that the new GaussianBlur works appropriately.

I'm waiting for your feedback :)

File: test/filters/test_gaussian.py [hunk 12]
```diff
@@ -51,6 +51,36 @@ def test_separable(ksize_x, ksize_y, sigma, device, dtype):
     assert_close(out, out_sep)
 
 
+@pytest.mark.parametrize("window_size", [5, 11])
+@pytest.mark.parametrize("sigma", [torch.tensor([1.5, 5.0]), torch.tensor([1.5, 5.0])])
+def test_get_n_gaussian_kernel(window_size, sigma):
+    kernel = kornia.filters.get_gaussian_kernel1d(window_size, sigma)
+    assert kernel.shape == (2, window_size)
+    assert kernel.sum(1)[0].item() == pytest.approx(1.0)
```
Comment (edgarriba) [inline test/filters/test_gaussian.py:59]: Users assert_close from pytorch instead

Comment (juliendenize) [reply test/filters/test_gaussian.py]: should be good now

**Review by edgarriba — COMMENTED** [review]
Can you also add tests cases for sigma as torch tensor in each function we introduce the new type ? After this LGTM

**Comment by juliendenize** [comment]
> 

Sure, but could you tell me what functions you are thinking about, please? I believe it is done for the gaussian part, so are you speaking about other kinds of transforms? 

In such a case, I looked at for example elastic transform but it looks like I would have to also update the type hinting and accept sigma tensors (with batch size B ?)

**Review by edgarriba — APPROVED** [review]
Oh, you are right. For some reason I missed those test

**Comment by edgarriba** [comment]
Oh you are right, for some any idea about the typing issue?

**Comment by juliendenize** [comment]
Do you mean for the test that didn't pass or for using sigma as tensor such as in elastic transform?

For the former, I checked the list that didn't pass and I don't think there is stuff I've changed.

For the latter, I think the type hinting that should be used is:
`Union[torch.Tensor, Tuple[float, float]]`

However, I would probably suggest that transforms such as elastic be updated for gaussian blur to either accept sigma float or tensor, convert to tensor, and accept batch size 1 or B.

**Comment by juliendenize** [comment]
Hello,

By checking the docstring test that failed, I noticed that for elastic transform it is expected to have sometimes a `requires_grad=True` for `sigma`. I had to update the function accordingly so this is why I ask for another review of the code.

File: kornia/filters/filter.py [hunk 13]
```diff
@@ -181,7 +181,7 @@ def filter2d_separable(
...
```
Comment (edgarriba) [inline kornia/filters/filter.py:184]:
```suggestion
    out_x = filter2d(input, kernel_x.unsqueeze(-2), border_type, normalized, padding)
```

Comment (edgarriba) [reply kornia/filters/filter.py:184]: to follow same convention as y kernel

**Review by edgarriba — APPROVED** [review]
LGTM. @ducha-aiki see some failing tests. Maybe this pr needs rebase ?

**Comment by juliendenize** [comment]
> LGTM. @ducha-aiki see some failing tests. Maybe this pr needs rebase ?

For the failing doctest, there is an issue because sometimes it expects scientific notation sometimes it does not, which causes trouble, is there a way to test it locally to avoid multiple tests tryout?

**Comment by edgarriba** [comment]
you can explore this: https://github.com/kornia/kornia/blob/master/Makefile#L17

we have in the makefile several commands useful tor developers

**Comment by juliendenize** [comment]
> you can explore this: https://github.com/kornia/kornia/blob/master/Makefile#L17
> 
> we have in the makefile several commands useful tor developers

Thanks, I've been able to debug the doctest thanks to `make doctest`, I was always launching `make test` which wasn't testing everything.

You mentioned that @ducha-aiki had seen other bugs, but I don't know where, when I checked output fails from the different tests, I thought bugs were caused by other functions than the one I touched, except for doctest, but I might have missunderstood.

**Comment by juliendenize** [comment]
Hello @edgarriba, thanks for the tests, apparently the problem for the test-lts failures is the Union type hint that I used. Is there any recommendation or convention in Kornia to change them? 

For the test-typing I think the failures are not related to this PR, am I correct?

**Comment by edgarriba** [comment]
ok, lets get merged this to be includedin 0.6.8

**Comment by juliendenize** [comment]
Thanks for merging master in the branch. To fix lts tests, do you have a preference for changing the type hinting of the gaussian blur functions so that Union is discarded (meaning keeping either float or tensor I would assume)?

**Comment by edgarriba** [comment]
@juliendenize seems that this PR breaks support included in torch lts specifically for the Union stuff. @ducha-aiki  any thoughts here ? might be a bit critical to break backward compatibility with lts

**Comment by ducha-aiki** [comment]
@edgarriba I would prefer to not break compatability with LTS. Probably the best workaround would be to create one more GaussianBlurT with tensor sigma, or like that

**Comment by juliendenize** [comment]
Hello @edgarriba @ducha-aiki , following your latest remarks, I added new functions that take sigma tensor parameters to remove the `Union[torch.tensor, Tuple[float, float]]` issue.

The functions taking float actually convert sigma to tensor and then call the new functions. It has the following advantages:
- only one implementation to maintain
- possibility to have batch generation of kernels and gaussian.

There are no longer issues with tests for the LTS. However, there is some kind of duplicity of documentation between functions taking floats and tensors ,and I don't really know if it is an issue for you or not.

File: kornia/augmentation/_2d/intensity/gaussian_blur.py [hunk 14]
```diff
@@ -1,26 +1,31 @@
...
+from kornia.augmentation import random_generator as rg
 from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
 from kornia.constants import BorderType
-from kornia.filters import gaussian_blur2d
+from kornia.filters import gaussian_blur2d_t
 
 
 class RandomGaussianBlur(IntensityAugmentationBase2D):
-    r"""Apply gaussian blur given tensor image or a batch of tensor images randomly.
+    r"""Apply gaussian blur given tensor image or a batch of tensor images randomly. The standard deviation is
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/gaussian_blur.py:13]:
```suggestion
    r"""Apply gaussian blur given tensor image or a batch of tensor images randomly.
    
    The standard deviation is sampled for each instance.
```

File: kornia/augmentation/_2d/intensity/gaussian_blur.py [hunk 15]
```diff
@@ -1,26 +1,31 @@
...
+    sampled for each instance.
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/gaussian_blur.py:14]:
```suggestion
```

File: kornia/augmentation/_2d/intensity/gaussian_blur.py [hunk 16]
```diff
@@ -52,17 +57,35 @@ def __init__(
...
+
+        if not silence_instantiation_warning:
+            warnings.warn(
+                "`RandomGaussianBlur` has changed its behavior and now randomly sample sigma for both axes. "
+                "To retrieve old behavior please consider using kornia.filters.GaussianBlur2d",
+                category=DeprecationWarning,
+            )
+
+        self.flags = dict(kernel_size=kernel_size, separable=separable, border_type=BorderType.get(border_type))
+        self._param_generator = cast(rg.RandomGaussianBlurGenerator, rg.RandomGaussianBlurGenerator(sigma))
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/gaussian_blur.py:79]: wondering if `rg.PlainUniformGenerator` would just work here ?  /cc @shijianjian 
https://github.com/kornia/kornia/pull/1955/files#diff-0060b44997ec0a817aac02ac03016e66196785ef15af18ff36156fc8184eb438R68

File: kornia/filters/kernels.py [hunk 17]
```diff
@@ -14,14 +16,20 @@ def normalize_kernel2d(input: torch.Tensor) -> torch.Tensor:
...
-    device, dtype = None, None
-    if isinstance(sigma, torch.Tensor):
-        device, dtype = sigma.device, sigma.dtype
-    x = torch.arange(window_size, device=device, dtype=dtype) - window_size // 2
+    sigma_t = torch.tensor(sigma)
+
+    return gaussian_t(window_size, sigma_t)
+
+
+def gaussian_t(window_size: int, sigma: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/filters/kernels.py:24]:
```suggestion
def gaussian_t(window_size: int, sigma: Tensor) -> Tensor:
```
`from kornia.core import Tensor`

Comment (edgarriba) [inline kornia/filters/kernels.py:24]: maybe we should add some docs too ?

Comment (edgarriba) [reply kornia/filters/kernels.py:24]: and type, shape checks

**Review by edgarriba — APPROVED** [review]
LGTM. Now looks really good and clean.
Added some comments for minor improvements that can be achieved later. 

/cc @ducha-aiki @shijianjian review and merge

**Review by ducha-aiki — APPROVED** [review]
Fix doct test and good to go

**Comment by edgarriba** [comment]
not sure if the doctest issue is because of a bad rebase ?

**Comment by juliendenize** [comment]
@edgarriba hello! I was wondering how things are supposed to go from now to make the pull request, do you want me to make the changes you suggested? I think the doctest thing is not something fixable from my side as it wasn't from my PR.

**Comment by edgarriba** [comment]
i suggest first to rebase from master, things are have evolved fast recently.

**Comment by jramapuram** [comment]
Any update on this? Blur seems to be causing a perf differntial vs. torchvision because of not sampling sigma.

**Comment by juliendenize** [comment]
> Any update on this? Blur seems to be causing a perf differntial vs. torchvision because of not sampling sigma.

I'll push rebased on master soon, I had stuff to do that prevented me to take time on this PR. In the meantime, before the pull request is accepted, you can use my forked version either from this pull request or from the `sce` branch that I used with kornia 0.6.3 for my own experiments if you don't need a recent kornia version.

**Comment by jramapuram** [comment]
Appreciate it @juliendenize  🙏

**Comment by juliendenize** [comment]
@edgarriba, hi I rebased on master few days ago, is it good to go ? The test failikg isn't from my part.

**Review by edgarriba — APPROVED** [review]
LGTM. @johnnv1 might have some suggestions for subsequent PR to handle better the function overloading. But lets merge as it is

File: kornia/filters/kernels.py [hunk 18]
```diff
@@ -15,15 +16,21 @@ def normalize_kernel2d(input: Tensor) -> Tensor:
...
-    if isinstance(sigma, Tensor):
-        device, dtype = sigma.device, sigma.dtype
-    x = torch.arange(window_size, device=device, dtype=dtype) - window_size // 2
+def gaussian(window_size: int, sigma: float) -> torch.Tensor:
+    sigma_t = torch.tensor(sigma)
+
+    return gaussian_t(window_size, sigma_t)
+
+
+def gaussian_t(window_size: int, sigma: torch.Tensor) -> torch.Tensor:
```
Comment (johnnv1) [inline kornia/filters/kernels.py:25]: This can be a unique function again in #2094 -- after we start drop jit (#2120)
